### PR TITLE
feat: attach volume to ecs

### PR DESCRIPTION
**Pull request by houpeng80**

### CODE REVIEW

File: builder/ecs/step_run_source_server.go [hunk 1]
```diff
@@ -197,6 +189,13 @@ func (s *StepRunSourceServer) Cleanup(state multistep.StateBag) {
 	}
 
 	serverID := s.serverID
+	ui.Say(fmt.Sprintf("Detacheing the volume..."))
```
Comment (ShiChangkuo) [inline builder/ecs/step_run_source_server.go:192]: please move this line into `detachServerVolume`

File: builder/ecs/run_config.go [hunk 2]
```diff
@@ -150,7 +170,13 @@ type RunConfig struct {
 
 type DataVolume struct {
 	// The data disk size in GB.
-	Size int `mapstructure:"volume_size" required:"true"`
+	Size int `mapstructure:"volume_size" required:"false"`
+	// The ID of the image.
+	DataImageId string `mapstructure:"data_image_id" required:"false"`
+	// The ID of the snapshot.
+	SnapshotId string `mapstructure:"snapshot_id" required:"false"`
+	// The ID of the volume.
```
Comment (ShiChangkuo) [inline builder/ecs/run_config.go:178]: how about using `The ID of an existing Volumes` ?

File: builder/ecs/step_attach_volume.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+		}
+		if len(specified) > 1 {
+			return nil, fmt.Errorf("only one of volume_id, volume_size, data_image_id, snapshot_id can be"+
+				"specified, but `%s` were specified", strings.Join(specified, ","))
+		}
+	}
+	return dataVolumeWraps, nil
+}
+
+func attachVolume(ui packer.Ui, state multistep.StateBag, ecsClient *ecs.EcsClient, serverId string, dataVolume DataVolume) error {
```
Comment (ShiChangkuo) [inline builder/ecs/step_attach_volume.go:143]: it's better to define the function as `attachDataVolumes`

File: builder/ecs/step_attach_volume.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	_, err = WaitForAttachVolumeJobSuccess(ui, state, ecsClient, jobID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func createVolumeByImageId(ui packer.Ui, state multistep.StateBag, imsClient *ims.ImsClient, evsClient *evs.EvsClient,
+	serverId string, dataVolume DataVolume) error {
+	ui.Say(fmt.Sprintf("Getting volume size by data image id..."))
```
Comment (ShiChangkuo) [inline builder/ecs/step_attach_volume.go:174]: it's unnecessary to print the message to console, please use log.Printf

File: builder/ecs/step_run_source_server.go (hunk 1 above)
Comment (houpeng80) [reply builder/ecs/step_run_source_server.go]: done

File: builder/ecs/run_config.go (hunk 2 above)
Comment (houpeng80) [reply builder/ecs/run_config.go]: done

File: builder/ecs/run_config.go [hunk 5]
```diff
@@ -150,7 +170,13 @@ type RunConfig struct {
 
 type DataVolume struct {
 	// The data disk size in GB.
-	Size int `mapstructure:"volume_size" required:"true"`
+	Size int `mapstructure:"volume_size" required:"false"`
+	// The ID of the image.
```
[the remark this answers is not in the document]
Comment (houpeng80) [reply builder/ecs/run_config.go]: done

File: builder/ecs/step_attach_volume.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+const (
+	VolumeId DataVolumeType = iota
+	Size
+	DataImageId
+	SnapshotId
+)
+
+type StepAttachVolume struct {
+	DataVolumes []DataVolume
+	Config      *Config
```
[the remark this answers is not in the document]
Comment (houpeng80) [reply builder/ecs/step_attach_volume.go]: done

File: builder/ecs/step_attach_volume.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+func parseDataVolumes(dataVolumes []DataVolume) ([]DataVolumeWrap, error) {
+	dataVolumeWraps := make([]DataVolumeWrap, 0, len(dataVolumes))
+	for _, dataVolume := range dataVolumes {
+		specified := make([]string, 0)
+		if dataVolume.Type == "" {
+			dataVolume.Type = "SSD"
+		}
+		if dataVolume.VolumeId != "" {
+			specified = append(specified, "volume_id")
+			dataVolumeWraps = append(dataVolumeWraps, DataVolumeWrap{dataVolume, VolumeId})
```
[the remark this answers is not in the document]
Comment (houpeng80) [reply builder/ecs/step_attach_volume.go]: done

File: builder/ecs/step_attach_volume.go (hunk 3 above)
Comment (houpeng80) [reply builder/ecs/step_attach_volume.go]: done

File: builder/ecs/step_attach_volume.go (hunk 4 above)
Comment (houpeng80) [reply builder/ecs/step_attach_volume.go]: done

**Pull request merged**